File: commons/hexagon/templates/kotlin/src/main/kotlin/com/phenotype/application/ExampleService.kt
package com.phenotype.application

import com.phenotype.domain.errors.EntityNotFoundError
import com.phenotype.domain.errors.ValidationError
import com.phenotype.domain.model.ExampleEntity
import com.phenotype.domain.ports.inbound.CreateInput
import com.phenotype.domain.ports.inbound.UpdateInput
import com.phenotype.domain.ports.inbound.UseCaseService
import com.phenotype.domain.ports.outbound.Repository

/**
 * Application service implementing the use case
 */
class ExampleService(
    private val repository: Repository
) : UseCaseService {

    override suspend fun create(input: CreateInput): ExampleEntity {
        if (input.name.isBlank()) {
            throw ValidationError("Name cannot be blank")
        }

        val entity = ExampleEntity.create(
            name = input.name,
            description = input.description
        )

        repository.save(entity)
        return entity
    }

    override suspend fun getById(id: String): ExampleEntity {
        return repository.findById(id)
            ?: throw EntityNotFoundError("ExampleEntity", id)
    }

    override suspend fun update(input: UpdateInput): ExampleEntity {
        if (input.id.isBlank()) {
            throw ValidationError("ID cannot be blank")
        }

        val entity = getById(input.id)
        val updated = entity
            .rename(input.name)
            .updateDescription(input.description)

        repository.save(updated)
        return updated
    }

    override suspend fun delete(id: String) {
        if (!repository.exists(id)) {
            throw EntityNotFoundError("ExampleEntity", id)
        }
        repository.delete(id)
    }

    override suspend fun list(): List<ExampleEntity> {
        return repository.findAll()
    }
}

File: commons/hexagon/templates/elixir/lib/hexagonal_{project_snake}/application/example_service.ex
defmodule {{ project_module }}.Application.ExampleService do
  @moduledoc """
  Application service for managing example entities.
  """

  alias {{ project_module }}.Domain.Entities.ExampleEntity
  alias {{ project_module }}.Domain.Ports.Outbound

  @behaviour {{ project_module }}.Domain.Ports.Inbound

  defstruct [:repository]

  @type t :: %__MODULE__{
          repository: module()
        }

  @doc """
  Creates a new service instance.
  """
  def new(repository) do
    %__MODULE__{repository: repository}
  end

  @impl true
  def create_example(service, id, name, description) do
    with {:ok, entity} <- ExampleEntity.create(id, name, description),
         {:ok, saved} <- Outbound.save(service.repository, entity) do
      {:ok, saved}
    end
  end

  @impl true
  def get_example(service, id) do
    Outbound.get_by_id(service.repository, id)
  end

  @impl true
  def update_example(service, id, attrs) do
    Outbound.update(service.repository, id, attrs)
  end

  @impl true
  def delete_example(service, id) do
    Outbound.delete(service.repository, id)
  end

  @impl true
  def list_examples(service) do
    Outbound.get_all(service.repository)
  end

  @impl true
  def activate_example(service, id) do
    with {:ok, entity} <- Outbound.get_by_id(service.repository, id),
         {:ok, activated} <- ExampleEntity.activate(entity),
         {:ok, saved} <- Outbound.save(service.repository, activated) do
      {:ok, saved}
    end
  end

  @impl true
  def complete_example(service, id) do
    with {:ok, entity} <- Outbound.get_by_id(service.repository, id),
         {:ok, completed} <- ExampleEntity.complete(entity),
         {:ok, saved} <- Outbound.save(service.repository, completed) do
      {:ok, saved}
    end
  end

  @impl true
  def archive_example(service, id) do
    with {:ok, entity} <- Outbound.get_by_id(service.repository, id),
         {:ok, archived} <- ExampleEntity.archive(entity),
         {:ok, saved} <- Outbound.save(service.repository, archived) do
      {:ok, saved}
    end
  end
end
